thunder/dynamo/utils.py

```diff
@@ -0,0 +1,301 @@
+    """
+
+    type: SplitReasonType
+    info: str | None
+    exception: Exception | None = None
+
+
+@dataclasses.dataclass
+class SubgraphInfo:
+    """
+    A dataclass containing information about a subgraph.
+
+    Attributes:
+        original_graph_module (torch.fx.GraphModule): The original graph module.
+        split_graph_module (torch.fx.GraphModule): Optional. The graph module for the split subgraph.
+        thunder_compiled_fns (list[Callable]): List of thunder optimized callables. This could be None if there the graph module was not supported by thunder. Look at the `split_reasons` for further information.
+        compiled_functions (list[CompiledFunction]): A list of compiled functions derived from the subgraph. This will be a list with one function in case the graph was not split.
+        split_reasons (list[SplitReason] | None): Optional list of reasons explaining why the subgraph was split. Present only if there are was a split.
+    """
+
+    original_graph_module: torch.fx.GraphModule
+    split_graph_module: torch.fx.GraphModule
+    thunder_compiled_fns: list[Callable]
+    submodule_to_compiled_functions: Mapping[torch.fx.GraphModule, CompiledFunction]
+    split_reasons: list | None = None
+
+
+def _concrete_shape(x):
+    """
+    Get the concrete shape for a FakeTensor if it has `torch.SymInt` in its shape.
+    """
+
+    def get_backed_value(s):
+        if isinstance(s, torch.SymInt):
+            return s.node.hint
+        # Value is already concrete.
+        return s
+
+    return tuple(map(get_backed_value, x.shape))
+
+
+def try_execute_thunder_symbol(thunder_symbol: "Symbol", node: torch.fx.Node) -> tuple[bool, SplitReason | None]:
+    """
+    Attempts to execute a given Thunder symbol within a tracing context, using proxies for the node's arguments.
+
+    This function operates within a Thunder tracing context to generate proxies for the provided node's arguments.
+    It then attempts to execute the Thunder symbol with these proxies. If any exceptions occur during proxy creation
+    or execution, it returns a tuple indicating failure and provides a `SplitReason` detailing the exception.
+
+    Args:
+        thunder_symbol (Symbol): The Thunder symbol to be executed. This is expected to be a callable that can
+            operate on proxy arguments.
+        node (torch.fx.Node): The Torch FX node whose arguments are to be proxied and passed to the Thunder symbol.
+
+    Returns:
+        tuple[bool, SplitReason | None]: A tuple where the first element is a boolean whether the execution passed or failed.
+            The second element is a `SplitReason` object if an error occurred, or `None` if the execution was successful.
+    """
+    import thunder
+    from thunder.core.trace import TraceCtx
+    from thunder.core.proxies import proxy
+
+    @thunder._with_cache_info_ctx
+    def _run_with_cache_info():
+
+        # We need cache info here as the default dtype and device support
```

Comment:
I think we should pursue this seperately after this PR. Wdyt?

Reply:
Yes, definitely a separate PR type of work. Can Thunder functions query PyTorch themselves for the current default type and device? If `thunder.new_ones` implementation is modified to do that we don't need to query the cache and pass it explicitly.